include/ppx/knob.h

```diff
@@ -0,0 +1,166 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/log.h"
+
+#include <functional>
+#include <iostream>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+// ---------------------------------------------------------------------------------------------
+// KnobType
+// ---------------------------------------------------------------------------------------------
+
+enum class KnobType
+{
+    Unknown,
+    Bool_Checkbox
+};
+
+std::ostream& operator<<(std::ostream& strm, KnobType kt);
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// KnobConfigs
+// ---------------------------------------------------------------------------------------------
+
+struct KnobConfig
+{
+    std::string flagName;
+    std::string flagDesc;
+    std::string displayName;
+    int         parentId;
+};
+
+// Available knob types for the developer to specify
+struct BoolCheckboxConfig : KnobConfig
+{
+    bool                      defaultValue;
+    std::function<void(bool)> callback;
+};
+
+// ---------------------------------------------------------------------------------------------
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+class Knob
+{
+public:
+    Knob() {}
+    Knob(const KnobConfig& config, KnobType type)
+        : displayName(config.displayName), flagName(config.flagName), flagDesc(config.flagDesc), type(type) {}
+    virtual ~Knob() {}
+
+    std::string        GetDisplayName() { return displayName; }
+    std::string        GetFlagName() { return flagName; }
+    std::string        GetFlagDesc() { return flagDesc; }
+    KnobType           GetType() { return type; }
+    void               AddChild(Knob* child) { children.push_back(child); }
+    std::vector<Knob*> GetChildren() { return children; }
+
+    virtual void Draw()  = 0;
+    virtual void Reset() = 0;
+
+    // Used for Bool_Checkbox
+    virtual bool GetBoolValue()
+    {
+        PPX_LOG_ERROR("Flag of type " << type << " does not support GetBoolValue()");
+        return false;
+    };
+    virtual void SetBoolValue(bool newVal, bool updateDefault = false)
+    {
+        PPX_LOG_ERROR("Flag of type " << type << " does not support SetBoolValue()");
+    };
+
+protected:
+    std::string        displayName;
+    std::string        flagName;
+    std::string        flagDesc;
+    KnobType           type;
+    std::vector<Knob*> children;
+};
+
+class KnobBoolCheckbox
```

Comment:
Naming seems to be following the UI you'll use. Shouldn't it be independant?
KnobBoolean
KnobLinearScale
KnobString
...

Reply:
That's fair. Although for UI items like dropdown, maybe the type could be part of the template?
KnobSlider<int> vs KnobSlider<float> by ex.

Reply:
edited the comment, github was hiding the `<>` part if not quoted